core/src/km_core_state_api.cpp

```diff
@@ -363,4 +364,32 @@ km_core_cp * km_core_state_context_debug(
   result[s.size()] = 0;
 
   return result;
-}
\ No newline at end of file
+}
+
+static bool
+state_has_action_type(km_core_state *state, uint8_t type) {
+  return std::any_of(
+      state->actions().begin(), state->actions().end(),
+        [type](const km::core::action &a) { return a.type == type; });
+}
+
+bool
+state_should_invalidate_context(km_core_state *state,
+                     km_core_virtual_key vk,
+                     uint16_t modifier_state,
+                     uint8_t is_key_down,
+                     uint16_t _kmn_unused(event_flags)) {
+  // if emit_keystroke is present, check if a context reset is needed
+  if (state_has_action_type(state, KM_CORE_IT_EMIT_KEYSTROKE)) {
+    if (vk == KM_CORE_VKEY_BKSP && state->context().empty()) {
+        return true; // context is empty - so pass back
```

Comment:
Is this test going to be before or after the context has been manipulated by the rule processor? That is, if we have a single char in context, then user presses backspace, could this result in a context reset when we don't want it? (Yes it may not cause trouble but I want us to be clear on our boundaries and order of execution)

Reply:
after

Reply:
that suggests that bksp for the first character will delete that character and _also_ pass it back to the app for a second deletion?

Reply:
Yes, that is correct, we can't rely on the state->context().empty(). I originally thought we could use  `code_points_to_delete` value from the action struct. However, the kmx processor currently only gives the action list.   This information could be found by checking each action item for KM_CORE_IT_BACK then `action_items->backspace.expected_type==KM_CORE_BT_CHAR`. 
I think there a few cleaner choices here.

1. Move  the new code we have in `km_core_state_api` starting at line 55 
```
 if (state_should_invalidate_context(state, vk, modifier_state, is_key_down, event_flags)) {
    state->context().clear();
```
To inside `state::apply_actions_and_merge_app_context()` after line   `action_item_list_to_actions_object(action_items, &this->_action_struct);`.  This way we can just look at the action struct for `code_points_to_delete`.  A bonus here is that this will also mean no change is needed when the kmx_processor is updated to use the action_struct instead of action_list.

2. Go with what we discussed as an option in or zoom meeting. Turn `km_core_bool emit_keystroke;` member of the action struct to a tri-state.  0 = do nothing; 1 = emit_key_stroke; 2 = emit_key_stroke and invalidate context. 
This puts the decision back on the processors though.

Reply:
@srl295 
Further thoughts looking again at the "kmx_processor" and the "ldml processor". All we have to do is remove the check for `state->context().empty(`).
In the case of a backspace key being pressed the processors only set  `KM_CORE_IT_EMIT_KEYSTROKE` when the context is empty.  Therefore that check has already been made. That is if action is  `KM_CORE_IT_EMIT_KEYSTROKE`  and `vkey== KM_CORE_VKEY_BKSP`. We do not need to check if the context is empty here. 

I think we can keep the change suggest in option 1 for the future when the action list is deprecated.